### PR TITLE
wip: gh-actions

### PR DESCRIPTION
**Pull request by d-chambers**
This PR explores using github actions for ObsPy's CI.

Currently it is only a proof of concept. Additional work/consensus  is needed before removing travis/appveyor. 

features:

1. Run tests on windows, linux, macos when a new PR is opened (or commits are pushed to the PR branch). We would still need to configure some of `obspy-runtest`'s command line arguments for the test report.

2. Lint code once as a dedicated job. Should replace the formatting tests in core.

3. Automate release to pypi when new release tag is created. We would need to generate a pypi token and add a github secret. 

4. Calculate coverage stats, update badge when commits added to master. This will need some additional configuration.

Thoughts?

### CODE REVIEW

**Comment by heavelock** [comment]
It looks like Actions are able to do everything I though to do in Azure. So I think it does not make much sense of trying Azure separately.

Apart of the processes you listed, I think it could be useful to add some on-tag triggered workflows like building docker images we are publishing on the DockerHub (see action https://github.com/marketplace/actions/build-tag-publish-docker ) as well as building packages for different distros (ie.: building with use of FPM https://github.com/bpicode/github-action-fpm or https://github.com/jrl-umi3218/github-actions )

**Comment by d-chambers** [comment]
If you want to test any of that or feel free to push to this branch

**Comment by heavelock** [comment]
I tested building images and pushing them to DockerHub on my fork and I have to admit I thought it will be more complicated that it is in reality. https://github.com/heavelock/obspy/pull/3 I managed to prepare a test workflow building few base images according to a build matrix. They are all pushed under proper name to a DockerHub repository.  See: https://hub.docker.com/repository/docker/heavelock/obspy-testing/tags

However it's just a first step since those baseimages are in the repository mostly for the testing purposes and do not even contain obspy itself, I would propose to make building baseimages on push-on-master trigger, followed by a running tests on them. I do not think it is that necessary to run tests on all of them on commit in PR hence the proposition of limiting it to the pushing on master only.

What is more, I would propose to extend our Dockerimages by an image that would be targeted to a wider audience and focused on providing a fully working obspy instalation to be downloaded from the hub. This could be done for tags, master branch as well as pull requests if we feel so. In this way potential users don't need to mess with their local environment to do quick prototyping or whatsoever. Recently I found out that Bioconda is providing a separate dockerimage for each of their packages and it was super useful for me when I failed to install some dependencies on Windows. 

What do you think? 
I ping also @megies.

**Comment by d-chambers** [comment]
I like the idea of automatically creating an obspy docker image for the tags. The docker tests are a great idea too!

Before we spend too much more time on it I think we should get a consensus on ObsPy's use of GH actions from other devs. 

We could:

:rocket: - use GH actions exclusively for CI, removing the other services
:heart: - run the current CI in parallel with GH actions, at least for awhile   
:+1: - use GH actions just for things the current CI doesn't do
:-1:  - don't use GH actions

Anyone care to vote?

**Comment by heavelock** [comment]
I am voting for running current CI with GH at the same time to see what are the maintenance costs of Actions against current setup. 

However, I am currently strongly for moving to the Actions. Having that tiny experience with Actions compared to other CI services I see one huge advantage of GHA which is integration with GitHub itself. It makes just super easy to run CI in your fork in case you want to test something and I guess it is also well integrated in case of tasks ran outside of PR interface (for example nightly master test or docker packaging).

**Comment by megies** [comment]
@heavelock @d-chambers I didn't follow everything to closely but from what I can see it makes sense, so as far as I'm concerned you folks go ahead with what you think is best.

@heavelock if I understood correctly you were proposing to run our docker tests through GH actions on pushed tags? That'd be awesome, to be honest, so far I was running them locally through my docker, I even had to use a separate external HDD since the base images build up quite a lot on disk space.

**Comment by megies** [comment]
@heavelock as far as Debian packaging is concerned, the current setup is quite iffy (we just got it to work somehow and always put new patches on top of the holes where water started comin in ;)), so while in principle it *should* work through GH actions, I doubt it'll ever really work that way, I usually have to tweak packaging for every new Debuntu distro.. so tinkering is pretty much always involved.

**Comment by megies** [comment]
Just a nooby question, so GH actions is running their own infrastructure and this will mean dropping Travis/appveyor and Circle completely?

**Comment by heavelock** [comment]
>  understood correctly you were proposing to run our docker tests through GH actions on pushed tags? 
Yup, that's what I thought about doing. 
Plus a second thing I thought about is preparing a new image/images that have obspy installed and would be there for anyone interested in quick checking out of obspy without messing around with their own environment.
This could be also extended to building an image per PR so reviewers could just "docker run" that precise version of obspy.

Regarding the debian packaging. Probably it's not the best idea to include that in CI since our rate of releasing new versions is not that high. 

> Just a nooby question, so GH actions is running their own infrastructure and this will mean dropping Travis/appveyor and Circle completely? 

GH is running actions on their own infrastructure and it's free for OSS.  It does not necessarily mean that we have to drop all of those CI providers but since GH Actions can take over all of their current usages, we can do that.

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+      - name: install obspy
+        shell: bash -l {0}
+        run: |
+          pip install -e .
+
+      - name: run test suite
+        shell: bash -l {0}
+        run: |
+          obspy-runtests
```
Comment (megies) [inline .github/workflows/coverage.yml:31]: there's more to add here, but I'm pretty sure you're aware and just wanted to keep it simple for the first tests :+1: 

```
  - if [[ "$TRAVIS_PULL_REQUEST" == "false" ]]; then
      coverage run --rcfile=.coveragerc --source=${MODULELIST} -m obspy.scripts.runtests --no-flake8 --all -n travis-ci -r --ci-url="https://travis-ci.org/${TRAVIS_REPO_SLUG}/jobs/${TRAVIS_JOB_ID}";
    else
      coverage run --rcfile=.coveragerc --source=${MODULELIST} -m obspy.scripts.runtests --no-flake8 -n travis-ci -r --ci-url="https://travis-ci.org/${TRAVIS_REPO_SLUG}/jobs/${TRAVIS_JOB_ID}" --pr-url="https://github.com/obspy/obspy/pull/${TRAVIS_PULL_REQUEST}" $MODULELISTSPACES;
```

**Comment by megies** [comment]
>  Plus a second thing I thought about is preparing a new image/images that have obspy installed and would be there for anyone interested in quick checking out of obspy without messing around with their own environment.

That might be for a very select tech-savvy minority, so basically people that probably have no issues installing from master branch.. but maybe it could be interesting in combination with mybinder.org, basically preparing a fully setup static base image that can be run in mybinder.org, similar to what we do for seismo-live currently (https://github.com/krischer/seismo_live/tree/master/docker_baseimage)

> Regarding the debian packaging. Then it's probably not the best idea to include that in CI since our rate of releasing new versions is not that high to have significant time saving here.

I agree, it's a nice thought, but I don't think it'd work

> GH is running actions on their own infrastructure and it's free for OSS. It does not necessarily mean that we have to drop all of those CI providers but since GH Actions can take over all of their current usages, we can do that.

I didn't look at the speeds or if we get queued up a lot, but I guess we can just switch and have everything in one place if it works sufficiently well :+1:

**Comment by heavelock** [comment]
> 
> 
> > Plus a second thing I thought about is preparing a new image/images that have obspy installed and would be there for anyone interested in quick checking out of obspy without messing around with their own environment.
> 
> That might be for a very select tech-savvy minority, so basically people that probably have no issues installing from master branch.. but maybe it could be interesting in combination with mybinder.org, basically preparing a fully setup static base image that can be run in mybinder.org, similar to what we do for seismo-live currently (https://github.com/krischer/seismo_live/tree/master/docker_baseimage)
> 
> > Regarding the debian packaging. Then it's probably not the best idea to include that in CI since our rate of releasing new versions is not that high to have significant time saving here.
> 
> I agree, it's a nice thought, but I don't think it'd work
> 
> > GH is running actions on their own infrastructure and it's free for OSS. It does not necessarily mean that we have to drop all of those CI providers but since GH Actions can take over all of their current usages, we can do that.
> 
>That might be for a very select tech-savvy minority, so basically people that probably have no issues installing from master branch.. 

In this case I am not talking about images with bleeding edge version but with tagged ones. Just to be able to prototype something quickly with fresh env, on any platform. Just as an anecdotal example, I was recently working with some biological data on windows machine. It was simpler for me to install docker and run an image prepared by bioconda (kinda conda-forge organism but for biology) than installing the package with conda due to some inconsistencies in the dependencies.

File: .github/workflows/coverage.yml (hunk 1 above)
Comment (d-chambers) [reply .github/workflows/coverage.yml:31]: Yes sir. Hopefully GHA gives us the the PR URL, need to look into it more.

**Comment by megies** [comment]
Other jobs that come to mind are the docs builds (master and PRs). No idea how much resources we can use there but it'd be awesome to integrate those.

**Comment by heavelock** [comment]
Oh, I didn't know that the docbuilder is to be modified. I will take a look on that.

**Comment by d-chambers** [comment]
Update:

I almost have everything figured out. I needed to rebase on master and squash a few redundant commits. Am I going to mess with anyone if I force push?

I also had to modify the `obspy_github_api` package somewhat, but hopefully I improved it. [Here is the modified branch on my fork](https://github.com/d-chambers/obspy_github_api/tree/gh-actions).

The last issue I am having is that the test reports fail to upload. I am getting a "bad gateway error" which I think has something to do with the size of the request url but I am not sure. 

Here is the current state: https://github.com/d-chambers/obspy/pull/1

**Comment by heavelock** [comment]
> Am I going to mess with anyone if I force push?
I'm also working on my fork so don't worry about me.

**Comment by megies** [comment]
@heavelock i can help out with making sense of the docs building workflow if needed, it's just at some point we have to do something about it, docs building is running on an ancient EOL virtual 
machine

**Comment by megies** [comment]
> The last issue I am having is that the test reports fail to upload. I am getting a "bad gateway error" which I think has something to do with the size of the request url but I am not sure.

I've seen problems when there's too many errors and the traceback logs exceed some size. We might have to cut of tracebacks if there's too many, dunno @d-chambers

**Comment by barsch** [comment]
tests.obspy.org was down - has been fixed

**Comment by d-chambers** [comment]
> tests.obspy.org was down

That would do it :wink:.

Everything seems to be working now.

@heavelock, we may want to come to a consensus on how we ought to parse PR comments for configuring the CI. Currently I am using a modified branch of github.com/obspy/obspy_github_api which writes a small json file based on the magic strings it finds then uploads that as build artifact so down-stream jobs can use it. This means it only has to be done once per build. You might do something similar and just tack on a build docs job which gets skipped unless the appropriate flag is found in the json config.

File: .github/workflows/runtests.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      - uses: actions/checkout@v1
+
+      - uses: actions/setup-python@v1
+        with:
+          python-version: '3.x'
+
+      - name: create ci config
+        shell: bash -l {0}
+        run: |
+          pip install git+https://github.com/d-chambers/obspy_github_api@gh-actions
```
Comment (heavelock) [inline .github/workflows/runtests.yml:27]: I guess you want in the end to point here to the original repo?

**Comment by heavelock** [comment]
It's ok for me. It will be very useful to make docbuilder transition seamless.

File: .github/workflows/runtests.yml (hunk 2 above)
Comment (d-chambers) [reply .github/workflows/runtests.yml:27]: Yes, but I don't have permission to write to it. I will submit a PR then we can change it.

**Comment by d-chambers** [comment]
Well this is frustrating, now we are getting a ton of matplotlib image comparison failures :disappointed:

**Comment by megies** [comment]
@d-chambers image comparison issues have always been a recurring problem. What matplotlib was used when they popped up? We did bump mpl to 3.2.0 on master, so *all* images should be recreated with that version, maybe that helps? Any new ideas for how to improve image testing better/more robust are always welcome too.. :-)

**Comment by megies** [comment]
@d-chambers @heavelock our docs building broke down, because it's a Python 3.4 environment and it's reporting syntax errors on f-strings now. So, I'd have to setup a new environment with 3.7 and working versions of Sphinx etc.. that might be really time consuming, but I guess we'd need those exact version specs anyway to get it to run on gh actions? Does gh actions have a way to provide pre-built built stuff, like a zipped anaconda for docs building that is reused on every build? Because we only want to get a working environment once and then reuse it as long as we can because it's super annoying to get that environment right..

**Comment by d-chambers** [comment]
Hey @megies,

It would be easy to create an anaconda environment for building the docs. Notice how we are using [a test environment](https://github.com/obspy/obspy/blob/gh-actions/.github/test_conda_env.yml) already.

**Comment by d-chambers** [comment]
> Any new ideas for how to improve image testing better/more robust are always welcome too.

IMHO we should consider removing the image tests because they are difficult to maintain and their failures are usually meaningless. We should still test that the code runs and produces *something*, but not check the content. If an image does get messed up it is unlikely it would get included in a publication or something without human scrutiny (at which point our diligent users would open an issue).

Is there an easy way to regenerate all of the images used for the comparisons?

**Comment by megies** [comment]
> IMHO we should consider removing the image tests because they are difficult to maintain and their failures are usually meaningless

Maybe some in-between way would be best, that is reduce the number of actual image tests to a minimum (like one image per really important plot), and do most plot testing on the matplotlib objects without rendering them. (like [here](https://github.com/obspy/obspy/pull/2455/files#diff-d7555df0057d85cf2b46e78af6d5f883R564))

> If an image does get messed up it is unlikely it would get included in a publication or something without human scrutiny

Well, there are mess-ups and there are mess-ups. Errors come in all variants and some might be easy to spot (like a plot being white on black instead of black on white) but some might not be.. 

> Is there an easy way to regenerate all of the images used for the comparisons?

`obspy-runtests --keep-images` should do the trick, then its only about moving the images from the `testrun` folder one directory down to replace the baselines.

> It would be easy to create an anaconda environment for building the docs. Notice how we are using [a test environment](https://github.com/obspy/obspy/blob/gh-actions/.github/test_conda_env.yml) already.

Yes, the tricky thing is just finding an environment with version numbers that work. Our docs build is pretty hackish in some parts (patched plot directives etc) and Sphinx and other packages aren't very happy about it when their version numbers change. Basically, somebody would have to play around with setting up a fresh Python 3.7 anaconda env with some version numbers fixed (sphinx, mpl, ...) and checking if the docs build OK. This environment should then be frozen somehow. We definitly want to avoid having the dependencies resolved in every CI run, but should try to zip it and unzip the ready environment in CI to make this stable.

**Comment by d-chambers** [comment]
Ok, it seems matplotlib was getting re-installed by obspy using pip. When I pin the version in the conda environment file the issues go away even though the matplotlib versions were identical. It must be some mpl dependency or configuration causing the issues, but I have successfully kicked the can down the road :smile:

> Our docs build is pretty hackish in some parts

Is it feasible to update/change the doc builds to be more robust? I know its probably easier said than done but that seems like the proper thing to do. Maybe @trichter already has it figured out :+1: 

I think the testing part of this PR is ready for review. Do we want to keep the pypi release or code-cov calculation automations or should I save them for a potential future PR?

**Comment by megies** [comment]
> Is it feasible to update/change the doc builds to be more robust? I know its probably easier said than done but that seems like the proper thing to do. Maybe @trichter already has it figured out +1

Sure porting to an up-to-date Sphinx and whatnot would be good. That will be a considerable amount of work though, pretty sure.

> pypi release or code-cov calculation automations or should I save them for a potential future PR?

Not sure what you mean about pypi. Regarding codecov.. it usually was reported from one of the Travis builds and then somehow integrated into the PR checks

**Comment by d-chambers** [comment]
> Not sure what you mean about pypi. Regarding codecov.

Apart from running the test suite I have included two other workflows:

.github/workflows/coverage.yml - This one calculates masters' coverage (using codecov) anytime it changes. I haven't adopted this to obspy at all yet though so it will need more work if we want to use it.

.github/workflows/release_published.yml - This one publishes to pypi whenever you add a new version (tag). It will require generating a [pypi api token](http://pyfound.blogspot.com/2019/07/pypi-now-supports-uploading-via-api.html) and adding it to [github secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets).

I think I will remove both these for now and just focus on getting the test suite workflow finished. I believe there are only two more items before it is ready to merge:

1) merge my branch of obspy_github_api back into obspy/obspy_github_api and update the workflow accordingly.

2) Setup workflow to report coverage after running the test matrix.

Anything I am missing?

**Comment by megies** [comment]
> .github/workflows/release_published.yml - This one publishes to pypi whenever you add a new version (tag). It will require generating a [pypi api token](http://pyfound.blogspot.com/2019/07/pypi-now-supports-uploading-via-api.html) and adding it to [github secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets).

Sounds useful. it's not a big deal to do locally, but why not if it can be automated. will it work on pure tag pushes even if CI ran on the commit already, though? To me it always seems like git pushes commits (`git push`) and tags (`git push --tags`) separately and it's not quite clear what the implications of changing up that order are..



> Anything I am missing?

Sounds good. Building docs in gh-actions would be awesome as stated before, but thats a major work item I guess.. so can be postponed

**Comment by megies** [comment]
As pointed out by @ThomasLecocq, could be worth a look: https://github.blog/2020-06-18-introducing-github-super-linter-one-linter-to-rule-them-all/

**Comment by d-chambers** [comment]
@megies I think that looks really cool. It would be nice to lint everything in one go. However, I fear it will require a lot of change and configuration, so it may be best to do it in a different PR so that this one doesn't get too huge.

I will try to find some time to finish this soon. I just need to figure out how to correctly upload coverage info.

The other thing to address is [this PR](https://github.com/obspy/obspy_github_api/pull/2). Then we should be good to merge (and disable the other CI providers).

**Comment by megies** [comment]
@d-chambers @heavelock see #2662 
(also sorry for not being very active in here)

**Comment by d-chambers** [comment]
Ok,

I feel good about this now. Coverage is working (see [here](https://codecov.io/github/obspy/obspy/commit/69ce2fe50eb9249fb36acec86391a7181527e87d)) and obspy-github-api has been updated. Thanks to everyone who helped or provided input. 

IMHO its ready for merging.

**Comment by megies** [comment]
I'm guessing this will have the same `+TESTS:...` logic in comments? But I guess we can verify that later on

**Comment by d-chambers** [comment]
> I'm guessing this will have the same +TESTS:... logic in comments? But I guess we can verify that later on

Yes, it should have the same behavior as before using [obspy's github api](https://github.com/obspy/obspy_github_api). The recent work I did over there was mainly to make things a little easier to use, it did not remove any major functionality. 

This [job](https://github.com/obspy/obspy/blob/69ce2fe50eb9249fb36acec86391a7181527e87d/.github/workflows/runtests.yml#L14-L36) first parses the github PR looking for the magic comments. Then it uploads the interpreted configuration as a build artifact. Later jobs download the configuration file and pass its state to the coverage command [here](https://github.com/obspy/obspy/blob/69ce2fe50eb9249fb36acec86391a7181527e87d/.github/workflows/runtests.yml#L92-L98).

**Comment by megies** [comment]
Looks like there's some issue with gathering the coverage data?
Screenshot from 2020-07-07 10-00-08

**Comment by d-chambers** [comment]
> I'm guessing this will have the same +TESTS:... logic in comments? But I guess we can verify that later on

Funny thing here. Because of this comment coverage is failing. Obspy-github-api thinks there should be a module named `obspy....`. This exposes a bug though as the selectable modules should be restricted to actual obspy modules. I will push out a quick fix.

**Comment by megies** [comment]
omg, yeah funny bug then xD

**Comment by QuLogic** [comment]
Circle builds head of the PR branch, not a merge, so there's no config for it on this branch.

**Comment by d-chambers** [comment]
> Circle builds head of the PR branch, not a merge, so there's no config for it on this branch.

So we need to flip a switch on the circle CI dashboard to turn it off when we are ready?

**Comment by megies** [comment]
> So we need to flip a switch on the circle CI dashboard to turn it off when we are ready?

We'll see. I'll merge this now, in any case

**Comment by megies** [comment]
FYI matplotlib is building docs in CI nowadays: https://twitter.com/matplotlib/status/1283106646802268160
But I'll look into getting it up and running on our new server one of these days maybe

**Comment by megies** [comment]
@d-chambers gh actions is failing since two weeks, looks like they might have changed something about conda env specification? Added a new mandatory file?

Screenshot from 2020-07-30 15-21-07
Screenshot from 2020-07-30 15-20-39

**Comment by megies** [comment]
Is `validate` workflow even running on pushes to master @d-chambers ? Already when this PR was merged `validate` did not run..

Screenshot from 2020-07-30 15-39-13

**Pull request merged**